service/resource/certificate/update.go

```diff
@@ -42,24 +41,26 @@ func (r *Resource) ProcessUpdateState(ctx context.Context, obj, updateState inte
 		return microerror.Mask(err)
 	}
 
-	for _, certificateFile := range updateCertificateFiles {
-		r.logger.Log("debug", fmt.Sprintf("writing certificate: %s", certificateFile.path))
-		if err := afero.WriteFile(r.fs, certificateFile.path, []byte(certificateFile.data), r.certificatePermission); err != nil {
-			return microerror.Mask(err)
+	if updateCertificateFiles != nil {
+		for _, certificateFile := range *updateCertificateFiles {
+			r.logger.Log("debug", fmt.Sprintf("writing certificate: %s", certificateFile.path))
+			if err := afero.WriteFile(r.fs, certificateFile.path, []byte(certificateFile.data), r.certificatePermission); err != nil {
+				return microerror.Mask(err)
+			}
 		}
 	}
 
 	return nil
 }
 
-func toCertificateFiles(v interface{}) ([]certificateFile, error) {
+func toCertificateFiles(v interface{}) (*[]certificateFile, error) {
 	if v == nil {
 		return nil, nil
 	}
 
-	certificateFiles, ok := v.([]certificateFile)
+	certificateFiles, ok := v.(*[]certificateFile)
```

Comment:
This was the major issue here. We weren't being consistent with what we were passing around. Fixed everything to `*[]certificateFile`

Reply:
You are taking the value of a slice which is a pointer. A slice is a pointer by nature so there is no reason I can think of right now to deal with a pointer of a slice, because it already is a pointer. What you can do is working with a slice which items are pointers. The notation would then be `[]*certificateFile`. We should only do this though in case certificate files contain state that is supposed to be mutated during its lifetime. This is usually necessary when structures implement behaviour. I do not see this. I see a certificate file structure as stupid container to carry information around. So why all of this?

Reply:
Lets keep `[]certificateFile` and make its usage consistent that way.